src/Microsoft.FeatureManagement/FeatureManagerSnapshot.cs

```diff
@@ -24,53 +24,78 @@ public FeatureManagerSnapshot(IFeatureManager featureManager)
         public async IAsyncEnumerable<string> GetFeatureNamesAsync()
         {
             if (_featureNames == null)
+            {
+                _featureNames = await CreateListAsync(_featureManager).ConfigureAwait(false);
+            }
+
+            foreach (string featureName in _featureNames)
+            {
+                yield return featureName;
+            }
+
+            static async ValueTask<List<string>> CreateListAsync(IFeatureManager featureManager)
             {
                 var featureNames = new List<string>();
 
-                await foreach (string featureName in _featureManager.GetFeatureNamesAsync().ConfigureAwait(false))
+                await foreach (string featureName in featureManager.GetFeatureNamesAsync().ConfigureAwait(false))
                 {
                     featureNames.Add(featureName);
                 }
 
-                _featureNames = featureNames;
-            }
-
-            foreach (string featureName in _featureNames)
-            {
-                yield return featureName;
+                return featureNames;
             }
         }
 
-        public async Task<bool> IsEnabledAsync(string feature)
+        public Task<bool> IsEnabledAsync(string feature)
         {
-            //
-            // First, check local cache
-            if (_flagCache.ContainsKey(feature))
+            if (_flagCache.TryGetValue(feature, out Task<bool> task))
             {
-                return _flagCache[feature];
+                return task;
             }
 
-            bool enabled = await _featureManager.IsEnabledAsync(feature).ConfigureAwait(false);
+            return Core();
+
+            Task<bool> Core()
+            {
+                lock (_flagCache)
+                {
+                    if (_flagCache.TryGetValue(feature, out task))
+                    {
+                        return task;
+                    }
 
-            _flagCache[feature] = enabled;
+                    task = _featureManager.IsEnabledAsync(feature);
+                    _flagCache.Add(feature, task);
 
-            return enabled;
+                    return task;
+                }
+            }
         }
 
-        public async Task<bool> IsEnabledAsync<TContext>(string feature, TContext context)
+        public Task<bool> IsEnabledAsync<TContext>(string feature, TContext context)
         {
-            //
-            // First, check local cache
-            if (_flagCache.ContainsKey(feature))
+            if (_flagCache.TryGetValue(feature, out Task<bool> task))
             {
-                return _flagCache[feature];
+                return task;
             }
 
-            bool enabled = await _featureManager.IsEnabledAsync(feature, context).ConfigureAwait(false);
+            return Core();
 
-            _flagCache[feature] = enabled;
+            Task<bool> Core()
```

Comment:
This looks like an abuse of inline functions. The implementation could be taken out and replace the `return Core();` line with no resulting code duplication.

Reply:
This is outdated with 24faebd55c9ccd91a5f764dd90d9824dbbe09cc0.